src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs

```diff
@@ -227,6 +228,45 @@ public ForEachSymbols GetForEachSymbols(SemanticModel semanticModel, SyntaxNode
             }
         }
 
+        public ImmutableArray<IMethodSymbol> GetDeconstructionAssignmentMethods(SemanticModel semanticModel, SyntaxNode node)
+        {
+            if (node is AssignmentExpressionSyntax assignment && assignment.IsDeconstruction())
+            {
+                var builder = ArrayBuilder<IMethodSymbol>.GetInstance();
+                FlattenDeconstructionMethods(semanticModel.GetDeconstructionInfo(assignment), builder);
+                return builder.ToImmutableAndFree();
+            }
+
+            return ImmutableArray<IMethodSymbol>.Empty;
+        }
+
+        public ImmutableArray<IMethodSymbol> GetDeconstructionForEachMethods(SemanticModel semanticModel, SyntaxNode node)
+        {
+            if (node is ForEachVariableStatementSyntax @foreach)
+            {
+                var builder = ArrayBuilder<IMethodSymbol>.GetInstance();
+                FlattenDeconstructionMethods(semanticModel.GetDeconstructionInfo(@foreach), builder);
+                return builder.ToImmutableAndFree();
+            }
+
+            return ImmutableArray<IMethodSymbol>.Empty;
+        }
+
+        private static void FlattenDeconstructionMethods(DeconstructionInfo deconstruction, ArrayBuilder<IMethodSymbol> builder)
+        {
+            var method = deconstruction.Method;
+            if (method != null)
+            {
+                builder.Add(method);
+                return;
+            }
+
+            foreach (var nested in deconstruction.Nested)
```

Comment:
so just to be sure.  You can' thave both .Method and .Nested available at the same time?

Reply:
Good catch. There is a bug here. Actually, if there is a `.Method`, there *will* be a `.Nested`.